### ISSUE TITLE
my py of TC

### ISSUE THREAD

**Issue by tnvrbabu** [tnvrbabu/DS_Piazza#2]
from __future__ import print_function

#To display the log details on standard out of the system.
import logging
from optparse import OptionParser
import sys
import os
import re
import pandas as pd
import nltk
from time import time
import matplotlib.pyplot as plt

import numpy as np
from cStringIO import StringIO
from pyth.plugins.rtf15.reader import Rtf15Reader
from pyth.plugins.plaintext.writer import PlaintextWriter
from pyth.plugins.python.reader import PythonReader
import csv

#Build different models using scikit-learn 
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.feature_selection import SelectKBest, chi2
from sklearn.linear_model import RidgeClassifier
from sklearn.pipeline import Pipeline
from sklearn.svm import LinearSVC
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import Perceptron
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.naive_bayes import BernoulliNB, MultinomialNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neighbors import NearestCentroid
from sklearn.ensemble import RandomForestClassifier
from sklearn.utils.extmath import density
from sklearn import metrics

# Display progress logs on stdout
logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s %(levelname)s %(message)s')

# parse commandline arguments
op = OptionParser()
op.add_option("--report",
              action="store_true", dest="print_report",
              help="Print a detailed classification report.")
op.add_option("--chi2_select",
              action="store", type="int", dest="select_chi2",
              help="Select some number of features using a chi-squared test")
op.add_option("--confusion_matrix",
              action="store_true", dest="print_cm",
              help="Print the confusion matrix.")
op.add_option("--top10",
              action="store_true", dest="print_top10",
              help="Print ten most discriminative terms per class"
                   " for every classifier.")
#op.add_option("--all_categories",
#              action="store_true", dest="all_categories",
#              help="Whether to use all categories or not.")
op.add_option("--use_hashing",
              action="store_true",
              help="Use a hashing vectorizer.")
op.add_option("--n_features",
              action="store", type=int, default=2 ** 16,
              help="n_features when using the hashing vectorizer.")
#op.add_option("--filtered",
#              action="store_true",
#              help="Remove un necessary information that is easily overfit: "
#                   "headers, signatures, and quoting.")

(opts, args) = op.parse_args()
if len(args) > 0:
    op.error("this script takes no arguments.")
    sys.exit(1)

print(__doc__)
op.print_help()
print()

###############################################################################
# Load categories from the dataset (df)

os.chdir("C:/INSOFE/Main/Project")

df = pd.read_csv('TextClassification_Data.csv', delimiter=',')

#mytest = str(df["sub_categories"][1]).upper()

Sub_Category_New = []
for i in range(len(df)):
    Sub_Category_New.append(str(df["sub_categories"][i]).upper())

df.insert(7,'sub_category_new',Sub_Category_New)

ds_categories = df.sub_category_new.unique().tolist()

#ds_categories = []
#for x in Sub_Category_New:
#    if x not in ds_categories:
#        ds_categories.append(x)

#[i for i,x in enumerate(ds_categories) if x == Sub_Category_New[10]]

Sub_Category_Num = []
for i in range(len(Sub_Category_New)):
    Sub_Category_Num.append(int(str([j for j,x in enumerate(ds_categories) if x == Sub_Category_New[i]]).replace("[","").replace("]","")))

df.insert(8,'sub_category_num',Sub_Category_Num)

#df2 = df.sub_category_new
#df3 = df["sub_categories"]

#df = pd.DataFrame(df, Sub_Category_New, Sub_Category_Num)

# "fileid",	"SUMMARY", 	"DATA",	"categories", "sub_categories", 	"previous_appointment", 	"ID", "sub_categories_new", "sub_categories_num" 
#val =  str(df["fileid"][1]) + "," + str(df["SUMMARY"][1]) + "," + str(df["DATA"][1])+ "," + str(df["categories"][1]) + "," + str(df["sub_categories"][1]) + "," + str(df["previous_appointment"][1]) + "," + str(df["ID"][1]) + "," + str(Sub_Category_New[1]) + "," + str(Sub_Category_Num[1]).replace("[","").replace("]","")

#b = open('TextClassification_DataNew.csv', 'wb')
#a = csv.writer(b)
#val = ("fileid",	"SUMMARY", 	"DATA",	"categories", "sub_categories", 	"previous_appointment", 	"ID", "sub_categories_new", "sub_categories_num")
#a.writerow(val)
#for i in range(0, len(Sub_Category_Num)):
#    val =  str(df["fileid"][i]) + "," + str(df["SUMMARY"][i]) + "," + str(df["DATA"][i])+ "," + str(df["categories"][i]) + "," + str(df["sub_categories"][i]) + "," + str(df["previous_appointment"][i]) + "," + str(df["ID"][i]) + "," + str(Sub_Category_New[i]) + "," + str(Sub_Category_Num[i]).replace("[","").replace("]","")
#    a.writerows(val)
#b.close()
#
#df = pd.read_csv('TextClassification_DataNew.csv', delimiter=',')

## Or export it in many ways, e.g. a list of tuples
#tuples = [tuple(x) for x in df.values]
    
## or export it as a list of dicts
#dicts = df.to_dict().values()

dfa = df

print(len(df))

#For loop ceates a list of cleaned data for each row. Also it removes the data which 
#is not compatable with rtf format. And reset the index.
#So the length of given_data now will reset to the new value, verify its length as well. 
# Hence we can create a new column of parsed data and use that as out complete data.

df.replace('', np.nan, inplace=True)
df = df.dropna()
df = df.reset_index(drop=True)
df.head()

print(len(df))
#ds_new_cat = []
#df['sub_categories'] = df['sub_categories'].lower()

#ds_categories = df['sub_categories_new'].unique()
#for ds_category in ds_categories:
#    ds_new_cat.addlist(ds_category.lower)

print(len(ds_categories))
if(len(ds_categories)>0):
#    ls_categories = ds_categories.tolist()
#    [categories.lower() for categories in ls_categories]
    categories = ds_categories
else:
    categories = None
    
#if opts.all_categories:
#    categories = None
#else:
#    categories = [
#        'alt.atheism',
#        'talk.religion.misc',
#        'comp.graphics',
#        'sci.space',
#    ]

#if opts.filtered:
#    remove = ('headers', 'footers', 'quotes')
#else:
#    remove = ()

print("Loading all dataset for categories:")
print(categories if categories else "all")

# Split the random shuffled data between train & test

msk = np.random.rand(len(df)) < 0.8

data_train = df[msk]

data_test = df[~msk]

print(len(data_train))
print(len(data_test))

#data_train = fetch_20newsgroups(subset='train', categories=categories,
#                                shuffle=True, random_state=42,
#                                remove=remove)
#
#data_test = fetch_20newsgroups(subset='test', categories=categories,
#                               shuffle=True, random_state=42,
#                               remove=remove)

print('data loaded')

# target names should use only categories
target_names = categories

Train_data = data_train['SUMMARY']
Test_data = data_test['SUMMARY']

Train_data1 = Train_data.tolist()
Test_data1 = Test_data.tolist()

for i in range(len(Train_data1)):
    Train_data1[i] = re.sub(r'\s+', ' ', str(Train_data1[i]))
#    Train_data1[i] = re.sub('[\d]', '', Train_data1[i])

print(len(Train_data1))

for i in range(len(Test_data1)):
    Test_data1[i] = re.sub(r'\s+', ' ', str(Test_data1[i]))
#    Test_data1[i] = re.sub('[\d]', '', Test_data1[i])

print(len(Test_data1))

def size_mb(docs):
    return sum(len(s.encode('utf-8')) for s in docs) / 1e6

data_train_size_mb = size_mb(unicode(Train_data1))
data_test_size_mb = size_mb(unicode(Test_data1))

print("%d documents - %0.3fMB (training set)" % (
    len(data_train), data_train_size_mb))
print("%d documents - %0.3fMB (test set)" % (
    len(data_test), data_test_size_mb))
print("%d categories" % len(categories))
print()

# Build target values of a training set and a test set
y_train, y_test = data_train.sub_category_num, data_test.sub_category_num

print("Extracting features from the training data using a sparse vectorizer")

t0 = time()
if opts.use_hashing:
    print("Options uses hashing technique to process vectorizer")
    vectorizer = HashingVectorizer(stop_words='english', non_negative=True,
                                   n_features=opts.n_features)
    X_train = vectorizer.transform(Train_data1)
else:
    print("Options uses Tfidf technique to process vectorizer")
    vectorizer = TfidfVectorizer(sublinear_tf=True, max_df=0.5,
                                 stop_words='english')
    X_train = vectorizer.fit_transform(Train_data1)
duration = time() - t0

print("done in %fs at %0.3fMB/s" % (duration, data_train_size_mb / duration))
print("n_samples: %d, n_features: %d" % X_train.shape)

print()

print("Extracting features from the test data using the same vectorizer")
t0 = time()
X_test = vectorizer.transform(data_test["SUMMARY"])
duration = time() - t0
print("done in %fs at %0.3fMB/s" % (duration, data_test_size_mb / duration))
print("n_samples: %d, n_features: %d" % X_test.shape)
print()

# mapping from integer feature name to original token string
if opts.use_hashing:
    feature_names = None
else:
    feature_names = vectorizer.get_feature_names()

if opts.select_chi2:
    print("Extracting %d best features by a chi-squared test" %
          opts.select_chi2)
    t0 = time()
    ch2 = SelectKBest(chi2, k=opts.select_chi2)
    X_train = ch2.fit_transform(X_train, y_train)
    X_test = ch2.transform(X_test)
    if feature_names:
        # keep selected feature names
        feature_names = [feature_names[i] for i
                         in ch2.get_support(indices=True)]
    print("done in %fs" % (time() - t0))
    print()

if feature_names:
    feature_names = np.asarray(feature_names)

def trim(s):
    """Trim string to fit on terminal (assuming 80-column display)"""
    return s if len(s) <= 80 else s[:77] + "..."

###############################################################################
# Benchmark classifiers
def benchmark(clf):
    print('_' * 80)
    print("Training: ")
    print(clf)
    t0 = time()
    clf.fit(X_train, y_train)
    train_time = time() - t0
    print("train time: %0.3fs" % train_time)

    t0 = time()
    pred = clf.predict(X_test)
    test_time = time() - t0
    print("test time:  %0.3fs" % test_time)

    score = metrics.accuracy_score(y_test, pred)
    print("accuracy:   %0.3f" % score)

    if hasattr(clf, 'coef_'):
        print("dimensionality: %d" % clf.coef_.shape[1])
        print("density: %f" % density(clf.coef_))

        if opts.print_top10 and feature_names is not None:
            print("top 10 keywords per class:")
            for i, label in enumerate(target_names):
                top10 = np.argsort(clf.coef_[i])[-10:]
                print(trim("%s: %s" % (label, " ".join(feature_names[top10]))))
        print()

    if opts.print_report:
        print("classification report:")
        print(metrics.classification_report(y_test, pred,
                                            target_names=target_names))

    if opts.print_cm:
        print("confusion matrix:")
        print(metrics.confusion_matrix(y_test, pred))

    print()
    clf_descr = str(clf).split('(')[0]
    return clf_descr, score, train_time, test_time

results = []
for clf, name in (
        (RidgeClassifier(tol=1e-2, solver="lsqr"), "Ridge Classifier"),
        (Perceptron(n_iter=50), "Perceptron"),
        (PassiveAggressiveClassifier(n_iter=50), "Passive-Aggressive"),
        (KNeighborsClassifier(n_neighbors=10), "kNN"),
        (RandomForestClassifier(n_estimators=100), "Random forest")):
    print('=' * 80)
    print(name)
    results.append(benchmark(clf))

for penalty in ["l2", "l1"]:
    print('=' * 80)
    print("%s penalty" % penalty.upper())
    # Train Liblinear model
    results.append(benchmark(LinearSVC(loss='l2', penalty=penalty,
                                            dual=False, tol=1e-3)))

    # Train SGD model
    results.append(benchmark(SGDClassifier(alpha=.0001, n_iter=50,
                                           penalty=penalty)))

# Train SGD with Elastic Net penalty
print('=' * 80)
print("Elastic-Net penalty")
results.append(benchmark(SGDClassifier(alpha=.0001, n_iter=50,
                                       penalty="elasticnet")))

# Train NearestCentroid without threshold
print('=' * 80)
print("NearestCentroid (aka Rocchio classifier)")
results.append(benchmark(NearestCentroid()))

# Train sparse Naive Bayes classifiers
print('=' * 80)
print("Naive Bayes")
results.append(benchmark(MultinomialNB(alpha=.01)))
results.append(benchmark(BernoulliNB(alpha=.01)))

print('=' * 80)
print("LinearSVC with L1-based feature selection")
# The smaller C, the stronger the regularization.
# The more regularization, the more sparsity.
results.append(benchmark(Pipeline([
  ('feature_selection', LinearSVC(penalty="l1", dual=False, tol=1e-3)),
  ('classification', LinearSVC())
])))

# make some plots

indices = np.arange(len(results))

results = [[x[i] for x in results] for i in range(4)]

clf_names, score, training_time, test_time = results
training_time = np.array(training_time) / np.max(training_time)
test_time = np.array(test_time) / np.max(test_time)

plt.figure(figsize=(14, 8))
plt.title("Score")
plt.barh(indices, score, .2, label="score", color='navy')
plt.barh(indices + .3, training_time, .2, label="training time",
         color='c')
plt.barh(indices + .6, test_time, .2, label="test time", color='darkorange')
plt.yticks(())
plt.legend(loc='best')
plt.subplots_adjust(left=.25)
plt.subplots_adjust(top=.95)
plt.subplots_adjust(bottom=.05)

for i, c in zip(indices, clf_names):
    plt.text(-.3, i, c)

plt.show()
